mentat/code_file_manager.py

```diff
@@ -126,106 +121,66 @@ def _add_file(self, abs_path: Path):
-        changes = resolve_non_insertion_conflicts(changes, self.user_input_manager)
-        changes = resolve_insertion_conflicts(changes, self.user_input_manager, self)
-        if not changes:
-            return []
-
-        new_code_lines = self.file_lines[rel_path].copy()
-        if new_code_lines != self._read_file(rel_path):
-            logging.info(f"File '{rel_path}' changed while generating changes")
-            cprint(
-                f"File '{rel_path}' changed while generating; current file changes"
-                " will be erased. Continue?",
-                color="light_yellow",
-            )
-            if not self.user_input_manager.ask_yes_no(default_yes=False):
-                cprint(f"Not applying changes to file {rel_path}.")
-                return None
-
-        # Necessary in case the model needs to insert past the end of the file
-        last_line = len(new_code_lines) + 1
-        largest_changed_line = math.ceil(changes[0].last_changed_line)
-        if largest_changed_line > last_line:
-            new_code_lines += [""] * (largest_changed_line - last_line)
-
-        min_changed_line = largest_changed_line + 1
-        for change in changes:
-            if change.last_changed_line >= min_changed_line:
-                raise MentatError(f"Change line number overlap in file {change.file}")
-            min_changed_line = change.first_changed_line
-            new_code_lines = change.apply(new_code_lines)
-        return new_code_lines
-
-    def write_changes_to_files(self, code_changes: list[CodeChange]) -> None:
-        file_changes = defaultdict[Path, list[CodeChange]](list)
-        for code_change in code_changes:
-            rel_path = code_change.file
-            abs_path = self.config.git_root / rel_path
-            match code_change.action:
-                case CodeChangeAction.CreateFile:
-                    cprint(f"Creating new file {rel_path}", color="light_green")
-                    self._add_file(abs_path)
-                    with open(abs_path, "w") as f:
-                        f.write("\n".join(code_change.code_lines))
-                case CodeChangeAction.DeleteFile:
-                    self._handle_delete(code_change)
-                case CodeChangeAction.RenameFile:
-                    abs_new_path = self.config.git_root / code_change.name
-                    self._add_file(abs_new_path)
-                    code_lines = self.file_lines[rel_path]
-                    with open(abs_new_path, "w") as f:
-                        f.write("\n".join(code_lines))
-                    self._delete_file(abs_path)
-                    file_changes[code_change.name] += file_changes[rel_path]
-                    file_changes[rel_path] = []
-                    self.file_lines[code_change.name] = self._read_file(
-                        code_change.name
+    # Mainly does checks on if file is in context, file exists, file is unchanged, etc.
```

Comment:
We used to do these checks earlier, right? like when the change was created by parsing?

On the one hand, some these should be rare errors (creating a file that already exists, editing non-existent file). But editing a file not in context might be common, since we are starting to include parts of files in context. It's not ideal if Mentat crashes *after* the user has seen the changes and accepted them / spent time doing interactive selection or providing feedback.

Reply:
We still do do them in the current parser; but of course, we don't want to copy that code to every parser we make. I agree we should always do it in the parser though; what if we make it so that FileEdit does these checks on creation? That way it would be done at parsing for every parser. This also made me realize that we don't do any rename file checks here anymore; I'll fix that.

Reply:
Yes doing it on FileEdit creation sounds great!